Docs/5 research how to build a default payment link page

victorjosefcaro opened the pull request.

victorjosefcaro committed:
Create 001-paddle-default-payment-link-page.md

victorjosefcaro committed:
Clarified payment recovery under Notes section

victorjosefcaro committed:
Renamed en/researches to en/decisions

victorjosefcaro committed:
Changed 3 digits prefix to 4 digits prefix

Numbers00 commented:
Good evening,

1. Please add the source articles you used under Notes so I can review as well
2. The `separate route implementation` would be simpler in the long run. We're currently in the process of major refactors in the backend, in order to separate the endpoints for certain functionalities, since they're eating up 2k+ LoC and have many redundant middlewares adding overhead.
3. If I'm not wrong, I believe `Paddle-hosted recovery` is a feature provided by Retain, and is not an actual option for the implementation of the default payment page. Please review this option.
4. Please change the decision drivers to:
- Simple to implement and extend
- Future-proof

If you have any inquiries with (4), please reply here, thank you.

victorjosefcaro committed:
Update 0001-paddle-default-payment-link-page.md based on PR comments

- Updated Issue
- Updated Decision and Decision Outcome to implement separate route implementation
- Updated Consequences to align with decision
- Updated Decision Drivers based on PR comments
- Removed Paddle-hosted recovery in Considered Options and Pros and Cons
- Added documentation link for Payment Recovery in Notes

Numbers00 commented:
Please add a sequence diagram for each route via [sequencediagram.org](https://sequencediagram.org). Use these as reference:
* https://sequencediagram.org/instructions.html
* https://tinyurl.com/rikaiai-oidc-flow-1

Use these as entities:
1. User
2. Slack Frontend
3. Slack Backend
4. Paddle

The ticket is for how the default payment link page is to be built. I'm currently unsure how the flow would work. Take your time, you can start with one route then tell us in LINE when the diagram is ready. Thank you very much.

victorjosefcaro committed:
docs: rename markdown

victorjosefcaro committed:
docs: add sequence diagram draft

million23 commented on docs/en/decisions/0003-paddle-default-payment-link-page-implementation.md:
```diff
@@ -0,0 +1,77 @@
+
+## Considered Options
+
+- **Dynamic route implementation:** Implement a dynamic route that handles both new subscriptions and payment updates. In this case, a single route at `/checkout` would be implemented that determines its behavior based on URL parameters (e.g., `status`, `customerId`, `recoveryId`).
+- **Separate route implementation:** Implement separate routes for new subscriptions and payment updates. In this case, separate routes would be implemented like (e.g., `/checkout/new` and `/checkout/update`.)
```
These options can be used at the same time. 
For example, in the current setup of the frontend - it can be stated as `/app/store/checkout/$some_id` and `/app/store/checkout/new`

Notice that the route `/app/store/checkout/` has 2 sub-routes that can be dynamic or not

million23 commented on docs/en/decisions/0003-paddle-default-payment-link-page-implementation.md:
```diff
@@ -0,0 +1,77 @@
+
+## Decision Outcome
+
+We chose `separate route implementation`  because it provides clearer separation of concerns and simpler state management per route, aligning with the current backend refactoring efforts to separate endpoints for different functionalities.
```
If your concern is about state management - this is not the main concern

State management in React (especially in the current project structure) is now simpler than what was before

Leveraging URL params as state can reduce prop drilling (although, we dont really use useContext because of that same issue)

million23 commented on docs/en/decisions/0003-paddle-default-payment-link-page-implementation.md:
```diff
@@ -0,0 +1,77 @@
+
+| Option | Pros | Cons |
+| --- | --- | --- |
+| Dynamic route implementation | - Single entry point for all payment activities.<br>- Simplifies URL management in Paddle settings. | - More complex state management.<br>- Requires careful handling of different scenarios. |
+| Separate route implementation | - Clearer separation of concerns.<br>- Simpler state management per route. | - Multiple URLs to manage in Paddle settings.<br>- Requires consistent branding across routes. |
```
Please refer to the review 2 items above.

million23 reviewed (commented):
Please focus more ways/suggestions on building the payment link page rather than the routes of the payment link page

I am getting where this is headed but routes are not the main concern here.

victorjosefcaro commented:
Is it okay if I add a different MADR document that will focus more on the actual building of the payment link page or should I just edit the current MADR document to disregard the routing and focus on the building of the default payment link page

million23 commented:
@victorjosefcaro 
No need to create another MADR.
Edit the current document and add the common patterns of building the said page.
But do not disregard the previous entry - putting a <strike>strikethrough</strike> to them is enough

Your input in regards about routing and state management is stated well👍. You can practice that in your own projects while expanding the possibilities of what I stated in my comments.

Thank you very much.

victorjosefcaro committed:
docs: update markdown

Revised to focus more on building the default payment link page instead of the routing of the page.
